### PR TITLE
Tree attention

### PR DESCRIPTION
**Pull request by AmineDiro**
# Tree packing for AsyncGRPO

Adds `packing="tree"` to `AsyncGRPOConfig`. The default is unchanged.

Multi-turn / multi-generation rollouts contain a lot of repeated prefixes. Today those tokens are forwarded once per row:

```text
rollout 0:  [ prompt ] [ completion A ]
rollout 1:  [ prompt ] [ completion B ]
rollout 2:  [ prompt ] [ completion A ] [ later turn ]
            ^^^^^^^^^^
            forwarded again for every row
```

Tree packing turns rows from the same group into a prefix tree:

```text
                         +-- completion A -- later turn
prompt -----------------+
                         +-- completion B
```

and forwards the shared tree once:

```text
packed:
[ prompt | completion A | later turn | completion B ]
```

So shared tokens only go through the model once.

## Attention stays the same

The FlexAttention mask restores the original row semantics.

```text
                         A1 -> A2
P1 -> P2 -> P3 ----------+
                         B1 -> B2
```

`A2` can attend to:

```text
P1 P2 P3 A1 A2
```

but not:

```text
B1 B2
```

Likewise, `B2` sees its own path and not the `A` branch.

The tree mask reduces this to a simple ancestor check:

```text
q can see k  <=>  tin[k] <= tin[q] < tout[k]
```

## The loss is still defined on the original rollouts

Packing only deduplicates the forward pass. It does not deduplicate training targets.

The same packed hidden state may be used by several original rollouts:

```text
                     +--> target A, advantage a0
hidden[prompt_end] ---+--> target B, advantage a1
                     +--> target A, advantage a2
```

So the LM-head path now uses:

```text
pred_index -> which packed hidden state to read
target_id  -> which original token that state should predict
```

instead of a boolean training mask.

Every trained token keeps its own target, advantage, old log-probability, and rollout identity.

In fp32, the packed forward matches forwarding every row independently within:

- `9.5e-7` on log-probs
- `7.6e-6` on gradients

## The planner also accounts for sharing

With sequence packing, the cost of a sample is its full length.

With tree packing, the cost depends on what the row already contains:

```text
row already has:
    [ P1 P2 P3 A1 ]

new sample:
    [ P1 P2 P3 A1 A2 A3 ]

novel cost:
                A2 A3
                ^^^^^
                2 tokens
```

Each row is bounded by both:

```text
unique forwarded tokens <= token_budget
training loss terms     <= token_budget
```

The second bound matters because packing can keep forwarded tokens small while the number of training targets continues to grow.

## Measured

Qwen3-8B full fine-tuning over SWE-Gym rollouts, 4 FSDP2 ranks + 1 vLLM GPU, `token_budget=16384`, identical rollouts on both arms. Medians over a 20-step window.

|                           | sequence | tree     |
| ------------------------- | -------- | -------- |
| rollouts trained / second | 2.9      | **8.2**  |
| trained tokens / second   | 376      | **1297** |
| `batch/packing_ratio`     | 1.00     | **4.21** |
| rows per rank             | 1.7      | **5.7**  |
| `perf/fwd_bwd_s`          | 4.48s    | 5.56s    |

## A couple of things to keep in mind

**The individual step gets slower, and MFU goes down.**

That is expected: each step now trains on about `3.8x` more data, while MFU is computed from forwarded tokens, which are exactly the tokens packing removes.

The more useful metric here is:

```text
batch/trained_tokens_per_step / perf/fwd_bwd_s
```

**The gain also depends on `fork_threshold_tokens`.**

This is existing AsyncGRPO behavior and is unchanged by this PR.

A turn only becomes a separate training row when re-tokenizing it drifts by more than `fork_threshold_tokens`.

On the agentic rollouts measured here, the drift is around 40 tokens, while the default threshold is 1024. With the default, almost nothing forks and a long conversation tends to stay as one row.

In this setup:

```text
default threshold: packing ratio = 1.45
with forking:      packing ratio = 4.21
```

`rollout/fork_frac` tells you which regime a run is in.

## Implementation

The flow is roughly:

```text
rollout samples
     |
     v
planner
"which samples fit on each rank?"
     |
     v
TreePacking.pack
     |
     +--> prefix forest
     +--> packed input_ids
     +--> DFS stamps for attention
     +--> (pred_index, target_id) for the loss
     |
     v
model
     |
     +--> FlexAttention tree mask
     |
     v
hidden[pred_index]
     |
     v
LM head -> target_id -> GRPO loss
```

There is a new `trl/experimental/async_grpo/tree/` module for the tree packer and FlexAttention mask, plus `packing.py`, which puts sequence and tree packing behind the same protocol.

`patch_chunked_lm_head` now takes `(pred_index, target_id)` and gathers the hidden states that contribute to the loss. This is needed because one packed position can feed several loss terms, which a boolean mask cannot represent.

The planner, collator, and `compute_loss` all use the same flat contract, so they do not need to branch on the packing strategy.

There is no new dependency. The mask runs on Inductor's Triton template.

I also tried FlashAttention 4. It was about 7% faster on the training step, but its CuteDSL kernels and vLLM's FlashInfer both register types through `tvm-ffi`, and whichever loads second aborts. So a trainer and vLLM server sharing one environment could not both start reliably.

This PR also fixes `print_prompt_completions_sample` crashing on tool-call messages where `content=None`. This is unrelated to packing, so happy to split it out.

---

> [!NOTE]
> **Medium Risk**
> Refactors the experimental Async GRPO forward/loss path (LM head API, collator fields, attention backend) for every run, not only tree mode; incorrect packing or mask logic could change gradients while defaults stay `packing="sequence"`.
> 
> **Overview**
> Adds optional **`packing="tree"`** to Async GRPO so micro-batch rows can be **prefix forests** instead of concatenated sequences, deduplicating shared tokens (prompts, multi-turn prefixes) in the forward while preserving per-rollout causality via **FlexAttention** (`tree_flex`).
> 
> Introduces a **`PackingProtocol`** (`packing.py` + `tree/`) wired through the planner, collator, and trainer: tree mode uses **`PrefixForest`**, group-aware **`TokenBudgetBatcher`** placement (unique tokens + loss-term caps), and **`batch/packing_ratio`** metrics; row balancing moves from ΣL² to **packing-specific attention cost**.
> 
> **Training tensor contract changes for all packing modes:** batches drop per-token `completion_mask` in favor of flat **`pred_index` / `target_id` / `segment_id`**; **`patch_chunked_lm_head`** gathers hidden states at those indices so one packed position can feed multiple GRPO loss terms. Docs expand tree-packing ops guidance (throughput vs MFU, `fork_threshold_tokens`).
> 
> Also fixes **`print_prompt_completions_sample`** when assistant `content` is `None` (tool calls), with a small test.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 68361e786478a19f7bb84b9596174639a28992da. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/async_grpo/async_grpo_trainer.py [hunk 1]
```diff
@@ -433,36 +443,50 @@ class FixedCountBatcher(torch.utils.data.IterableDataset):
...
         self.microbatch_size = microbatch_size
+        self.packing = packing or SequencePacking()
 
     def __iter__(self):
         batch = []
         for sample in self.dataset:
             batch.append(sample)
             if len(batch) == self.microbatch_size:
-                yield _balance_by_squared_length(batch, self.num_processes)
+                yield _balance_by_attention_cost(self.packing.atoms(batch), self.num_processes, self.packing)
```
[the remark this answers is not in the document]
Comment (AmineDiro) [reply trl/experimental/async_grpo/async_grpo_trainer.py:476]: fixed in : [45577fc](https://github.com/huggingface/trl/pull/7328/commits/45577fcb6df1bbf92dcbc60150770cc0d08f4a0d)
